File: src/main/scala/org/dataalgorithms/chap01/scala/CustomPartitioner.scala
package org.dataalgorithms.chap01.scala

import org.apache.spark.Partitioner


/**
 * A custom partitioner
 * 
 * org.apache.spark.Partitioner:
 *           An abstract class that defines how the elements in a  
 *           key-value pair RDD are partitioned by key. Maps each 
 *           key to a partition ID, from 0 to numPartitions - 1.
 *           
 *
 * @author Gaurav Bhardwaj (gauravbhardwajemail@gmail.com)
 * 
 * @editor Mahmoud Parsian (mahmoud.parsian@yahoo.com)
 *
 */
class CustomPartitioner(partitions: Int) extends Partitioner {
  
  require(partitions > 0, s"Number of partitions ($partitions) cannot be negative.")

  def numPartitions: Int = partitions

  def getPartition(key: Any): Int = key match {
    case (k: String, v: Int) => math.abs(k.hashCode % numPartitions)
    case null                => 0
    case _                   => math.abs(key.hashCode % numPartitions)
  }

  override def equals(other: Any): Boolean = other match {
    case h: CustomPartitioner => h.numPartitions == numPartitions
    case _                    => false
  }

  override def hashCode: Int = numPartitions
}

File: src/main/java/org/dataalgorithms/chap01/spark/CustomPartitioner.java
package org.dataalgorithms.chap01.spark;

import org.apache.spark.Partitioner;
//
import scala.Tuple2;

/**
 * CustomPartitioner: a custom partitioner for Secondary Sort design pattern
 *
 * @author Gaurav Bhardwaj (gauravbhardwajemail@gmail.com)
 *
 * @editor Mahmoud Parsian (mahmoud.parsian@yahoo.com)
 *
 */
public class CustomPartitioner extends Partitioner {

    private static final long serialVersionUID = -4130944499909424595L;

    private final int numPartitions;

    public CustomPartitioner(int partitions) {
        assert (partitions > 0);
        this.numPartitions = partitions;
    }

    @Override
    public int getPartition(java.lang.Object key) {
        if (key == null) {
            return 0;
        } else if (key instanceof Tuple2) {
            @SuppressWarnings("unchecked")
            Tuple2<String, Integer> tuple2 = (Tuple2<String, Integer>) key;
            return Math.abs(tuple2._1.hashCode() % numPartitions);
        } else {
            return Math.abs(key.hashCode() % numPartitions);
        }
    }

    @Override
    public int numPartitions() {
        return numPartitions;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + numPartitions;
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        //
        if (obj == null) {
            return false;
        }
        //
        if (!(obj instanceof CustomPartitioner)) {
            return false;
        }
        //
        CustomPartitioner other = (CustomPartitioner) obj;
        if (numPartitions != other.numPartitions) {
            return false;
        }
        //
        return true;
    }
}
